<a href="https://colab.research.google.com/github/tentenvi/tteessstt/blob/main/lab1/%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5%E2%84%961.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Задание по практике 1

##Загрузка датасета
В файле classification datasets расположены ссылки на датасеты. Выберите один в соответствии со своим вариантом, скачайте и прочитайте файл

**Не забудьте загрузить нужные библиотеки**

Например: pip3 install pandas

In [4]:
import pandas as pd

In [5]:
df = pd.read_csv("/content/anime/anime.csv")
df

anime_id                                               name  \
0         32281                                     Kimi no Na wa.   
1          5114                   Fullmetal Alchemist: Brotherhood   
2         28977                                           Gintama°   
3          9253                                        Steins;Gate   
4          9969                                      Gintama&#039;   
...         ...                                                ...   
12289      9316       Toushindai My Lover: Minami tai Mecha-Minami   
12290      5543                                        Under World   
12291      5621                     Violence Gekiga David no Hoshi   
12292      6133  Violence Gekiga Shin David no Hoshi: Inma Dens...   
12293     26081                   Yasuji no Pornorama: Yacchimae!!   

                                                   genre   type episodes  \
0                   Drama, Romance, School, Supernatural  Movie        1   
1      Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64   
2      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
3                                       Sci-Fi, Thriller     TV       24   
4      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
...                                                  ...    ...      ...   
12289                                             Hentai    OVA        1   
12290                                             Hentai    OVA        1   
12291                                             Hentai    OVA        4   
12292                                             Hentai    OVA        1   
12293                                             Hentai  Movie        1   

       rating  members  
0        9.37   200630  
1        9.26   793665  
2        9.25   114262  
3        9.17   673572  
4        9.16   151266  
...       ...      ...  
12289    4.15      211  
12290    4.28      183  
12291    4.88      219  
12292    4.98      175  
12293    5.46      142  

[12294 rows x 7 columns]

##Первое знакомство с датасетом
В следующих ячейках выведите:


1.   Колонки датасета
2.   Первые 5 значений
3.   Последние 5 значений
4.   Статистику по числовым колонкам



In [7]:
df.columns

Index(['anime_id', 'name', 'genre', 'type', 'episodes', 'rating', 'members'], dtype='object')

In [8]:
df[:5]

anime_id                              name  \
0     32281                    Kimi no Na wa.   
1      5114  Fullmetal Alchemist: Brotherhood   
2     28977                          Gintama°   
3      9253                       Steins;Gate   
4      9969                     Gintama&#039;   

                                               genre   type episodes  rating  \
0               Drama, Romance, School, Supernatural  Movie        1    9.37   
1  Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64    9.26   
2  Action, Comedy, Historical, Parody, Samurai, S...     TV       51    9.25   
3                                   Sci-Fi, Thriller     TV       24    9.17   
4  Action, Comedy, Historical, Parody, Samurai, S...     TV       51    9.16   

   members  
0   200630  
1   793665  
2   114262  
3   673572  
4   151266

In [11]:

df.tail()

anime_id                                               name   genre  \
12289      9316       Toushindai My Lover: Minami tai Mecha-Minami  Hentai   
12290      5543                                        Under World  Hentai   
12291      5621                     Violence Gekiga David no Hoshi  Hentai   
12292      6133  Violence Gekiga Shin David no Hoshi: Inma Dens...  Hentai   
12293     26081                   Yasuji no Pornorama: Yacchimae!!  Hentai   

        type episodes  rating  members  
12289    OVA        1    4.15      211  
12290    OVA        1    4.28      183  
12291    OVA        4    4.88      219  
12292    OVA        1    4.98      175  
12293  Movie        1    5.46      142

In [13]:
df.describe(include="all")

anime_id                     name   genre   type episodes  \
count   12294.000000                    12294   12232  12269    12294   
unique           NaN                    12292    3264      6      187   
top              NaN  Shi Wan Ge Leng Xiaohua  Hentai     TV        1   
freq             NaN                        2     823   3787     5677   
mean    14058.221653                      NaN     NaN    NaN      NaN   
std     11455.294701                      NaN     NaN    NaN      NaN   
min         1.000000                      NaN     NaN    NaN      NaN   
25%      3484.250000                      NaN     NaN    NaN      NaN   
50%     10260.500000                      NaN     NaN    NaN      NaN   
75%     24794.500000                      NaN     NaN    NaN      NaN   
max     34527.000000                      NaN     NaN    NaN      NaN   

              rating       members  
count   12064.000000  1.229400e+04  
unique           NaN           NaN  
top              NaN           NaN  
freq             NaN           NaN  
mean        6.473902  1.807134e+04  
std         1.026746  5.482068e+04  
min         1.670000  5.000000e+00  
25%         5.880000  2.250000e+02  
50%         6.570000  1.550000e+03  
75%         7.180000  9.437000e+03  
max        10.000000  1.013917e+06

**Ответьте на следующие вопросы:**


1.   Какие средние значения в каждой числовой колонке вашего датасета?
2.   Минимальные и максимальные значения в каждой числовой колонке вашего датасета





Ответ 1: rating = 6.473902; members = 1.807134e+04

Ответ 2: минимальное = 1; максимальное = 34527


##Типизация
Присвойте соответствующим колонкам один из типов (category, float, int)

In [24]:
df.genre = df.genre.astype("category")
df.rating = df.rating.astype("Float64")
df.anime_id = df.anime_id.astype("Int64")

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12294 entries, 0 to 12293
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   anime_id  12294 non-null  Int64   
 1   name      12294 non-null  category
 2   genre     12232 non-null  category
 3   type      12269 non-null  category
 4   episodes  12294 non-null  object  
 5   rating    12064 non-null  Float64 
 6   members   12294 non-null  int64   
dtypes: Float64(1), Int64(1), category(3), int64(1), object(1)
memory usage: 977.2+ KB


Какие признаки вы можете выделить в своем датасете (категориальные, числовые и др., **минимум 3 разных**, помните что можно также выделить признаки более точно, например числовые - порядковые, количественные и т.д.)?

##Фильтрация
Офильтруйте 3 колонки вашего датасета и выведите результат

In [25]:
#Фильтр колонки 1
df[df["genre"]== "Supernatural"]

anime_id                                          name         genre  \
6069       9951                                     Jiu Se Lu  Supernatural   
7053      32191    Okitegami Kyouko no Bibouroku x Monogatari  Supernatural   
7134      11447                                     Yuureisen  Supernatural   
7699      23065                                        Kaidan  Supernatural   
7731       6832       Shoujouji no Tanuki-bayashi Ban Danemon  Supernatural   
7838       6831  Hinomaru Hatanosuke: Bakemonoyashiki no Maki  Supernatural   
7908       3930                         Issunboushi no Shusse  Supernatural   
8631      24677         Gegege no Kitarou: Kitarou Tanjou-hen  Supernatural   
9058      30921                                  Kacchikenee!  Supernatural   
9710      30217                                     Neo Faust  Supernatural   
9761       7495                Nonki na Tou-san Ryuuguu Mairi  Supernatural   
9851       7531                                    Onbu Obake  Supernatural   
10103      8685             Ryuuichi Manga Gekijou Onbu Obake  Supernatural   
10596      2366                           Touma Kishinden Oni  Supernatural   

          type episodes  rating  members  
6069     Movie        1    6.26      775  
7053       ONA        1    5.77     1771  
7134     Movie        1     5.7      609  
7699     Movie        1    5.13      293  
7731     Movie        1    5.09      628  
7838     Movie        1    4.89      593  
7908     Movie        1    4.69      685  
8631   Special        1    5.13      103  
9058     Movie        1    6.28      855  
9710   Special        1    5.38      138  
9761     Movie        1     5.5      105  
9851     Movie        1    6.08       97  
10103       TV       52    5.59      102  
10596       TV       25     6.4      303

In [28]:
#Фильтр колонки 2
df[df["rating"] >= 9.1]

anime_id                                               name  \
0         32281                                     Kimi no Na wa.   
1          5114                   Fullmetal Alchemist: Brotherhood   
2         28977                                           Gintama°   
3          9253                                        Steins;Gate   
4          9969                                      Gintama&#039;   
5         32935  Haikyuu!!: Karasuno Koukou VS Shiratorizawa Ga...   
6         11061                             Hunter x Hunter (2011)   
7           820                               Ginga Eiyuu Densetsu   
8         15335  Gintama Movie: Kanketsu-hen - Yorozuya yo Eien...   
9         15417                           Gintama&#039;: Enchousen   
9078      33607                                       Kahei no Umi   
9595      23005                                   Mogura no Motoro   
10400     30120                        Spoon-hime no Swing Kitchen   
10464     33662            Taka no Tsume 8: Yoshida-kun no X-Files   
10786     26313                    Yakusoku: Africa Mizu to Midori   

                                                   genre   type episodes  \
0                   Drama, Romance, School, Supernatural  Movie        1   
1      Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64   
2      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
3                                       Sci-Fi, Thriller     TV       24   
4      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
5                 Comedy, Drama, School, Shounen, Sports     TV       10   
6                Action, Adventure, Shounen, Super Power     TV      148   
7                         Drama, Military, Sci-Fi, Space    OVA      110   
8      Action, Comedy, Historical, Parody, Samurai, S...  Movie        1   
9      Action, Comedy, Historical, Parody, Samurai, S...     TV       13   
9078                                          Historical  Movie        1   
9595                                       Slice of Life  Movie        1   
10400                                    Adventure, Kids     TV  Unknown   
10464                                     Comedy, Parody  Movie        1   
10786                                        Drama, Kids    OVA        1   

       rating  members  
0        9.37   200630  
1        9.26   793665  
2        9.25   114262  
3        9.17   673572  
4        9.16   151266  
5        9.15    93351  
6        9.13   425855  
7        9.11    80679  
8         9.1    72534  
9        9.11    81109  
9078     9.33       44  
9595      9.5       62  
10400     9.6       47  
10464    10.0       13  
10786    9.25       53

In [27]:
#Фильтр колонки 3
df[df["type"] == "TV"]

anime_id                                               name  \
1          5114                   Fullmetal Alchemist: Brotherhood   
2         28977                                           Gintama°   
3          9253                                        Steins;Gate   
4          9969                                      Gintama&#039;   
5         32935  Haikyuu!!: Karasuno Koukou VS Shiratorizawa Ga...   
...         ...                                                ...   
11104     34522                          Wake Up, Girls! Shin Shou   
11106     34467                             Yami Shibai 4th Season   
11107     32615                                        Youjo Senki   
11110     34284  Yuuki Yuuna wa Yuusha de Aru: Washio Sumi no Shou   
11111     34445       Yuuki Yuuna wa Yuusha de Aru: Yuusha no Shou   

                                                   genre type episodes  \
1      Action, Adventure, Drama, Fantasy, Magic, Mili...   TV       64   
2      Action, Comedy, Historical, Parody, Samurai, S...   TV       51   
3                                       Sci-Fi, Thriller   TV       24   
4      Action, Comedy, Historical, Parody, Samurai, S...   TV       51   
5                 Comedy, Drama, School, Shounen, Sports   TV       10   
...                                                  ...  ...      ...   
11104                                       Drama, Music   TV  Unknown   
11106                     Dementia, Horror, Supernatural   TV  Unknown   
11107                                    Magic, Military   TV  Unknown   
11110               Drama, Fantasy, Magic, Slice of Life   TV        6   
11111               Drama, Fantasy, Magic, Slice of Life   TV        6   

       rating  members  
1        9.26   793665  
2        9.25   114262  
3        9.17   673572  
4        9.16   151266  
5        9.15    93351  
...       ...      ...  
11104    <NA>      381  
11106    <NA>     1838  
11107    <NA>     6652  
11110    <NA>     2593  
11111    <NA>     4439  

[3787 rows x 7 columns]

##Группировка
Сгруппируйте датасет по какому-либо категориальному признаку и выведите среднее значение численных колонок (см. метод mean в pandas)

In [30]:
#Группировка

#Группировка
group = df.groupby(["episodes"])
group.count()

anime_id  name  genre  type  rating  members
episodes                                              
1             5677  5677   5631  5677    5612     5677
10             114   114    114   114     112      114
100             10    10     10    10      10       10
1006             1     1      1     1       1        1
101              5     5      5     5       5        5
...            ...   ...    ...   ...     ...      ...
96               4     4      4     4       4        4
97               3     3      3     3       3        3
98               1     1      1     1       1        1
99               2     2      2     2       2        2
Unknown        340   340    329   315     188      340

[187 rows x 6 columns]

In [31]:
#Среднее значение
group.mean()

<ipython-input-31-8680372f086a>:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  group.mean()


anime_id    rating       members
episodes                                      
1         16154.635899  6.215451   7124.474546
10        17742.131579  6.566518  30331.947368
100            12066.9     6.563   1412.000000
1006           10241.0      5.92    153.000000
101             9461.2      7.34  19153.400000
...                ...       ...           ...
96             4643.25    6.8825   5911.250000
97              7223.0  7.176667  46635.333333
98              2268.0      7.67   5652.000000
99              7934.0      7.95  37393.000000
Unknown   30530.711765  6.067606   7621.344118

[187 rows x 3 columns]

##Задание для гениев *
Агрегируйте числовой признак по категориальному признаку с использованием метода **без использования** библиотеки pandas, и вычислите медиану и среднее значение для каждой категории.

In [ ]:
#Агрегация